source/SpinalHDL/Libraries/Pipeline/introduction.rst

```diff
@@ -0,0 +1,762 @@
+      }
+
+      // Let's connect those nodes by using registers stages and generate the related hardware
+      builder.genStagedPipeline()
+    }
+
+Stageable
+============
+
+Stageable class can be instanciated to represent some data which can go through the pipeline. Technicaly speaking, Stageable is a HardType which has a name and is used as a "key" to retrieve stuff.
+
+.. code-block:: scala
+    
+    val PC = Stageable(UInt(32 bits))
+    val PC_PLUS_4 = Stageable(UInt(32 bits))
+
+    val n0, n1 = Node()
+    val s01 = StageConnector(n0, n1)
+
+    n0(PC) := 0x42
+    n1(PC_PLUS_4) := n1(PC) + 4
+
+Note that I got used to name the Stageable instances using uppercase. This is to make it very explicit that the thing isn't a hardware signal, but are more like a "key" to access things.
+
+Node
+============
+
+Node mostly host the valid/ready arbitration signal, and the hardware signal required for all the Stageable values going through it.
+
+You can access its arbitration via :
+
+
+.. list-table::
+   :header-rows: 1
+   :widths: 1 5
+
+   * - API
+     - Description
+   * - node.valid
+     - Is the signal which specify if a transaction is present on the node
+   * - node.ready
+     - Is the signal which specify if the node transaction can move away.
+   * - node.isValid
+     - node.valid's read only accessor
+   * - node.isReady
+     - node.ready's read only accessor
+   * - node.isFiring
+     - True when the node transaction is successfuly moving futher (isValid && isReady && !isRemoved). Usefull to commit state changes
+   * - node.isMoving
+     - True when the node transaction is moving (isValid && (isReady || isRemoved)). Usefull to "reset" states
+   * - node.isRemoved
+     - True when the node is being flushed
+
+You can access its stageable's signals via : 
+
+.. list-table::
+   :header-rows: 1
+   :widths: 2 5
+
+   * - API
+     - Description
+   * - node(Stageable)
+     - Return the corresponding hardware signal
+   * - node(Stageable, Any)
+     - Same as above, but include a second argument which is used as a "secondary key". This ease the construction of multi lane hardware. For instance, when you have a multi issue CPU pipeline, you can use the lane Int id as secondary key
+   * - node.insert(Data)
+     - Return a new Stageable instance which is connected to the given Data hardware signal
+
+
+
+.. code-block:: scala
+    
+    val n0, n1 = Node()
+
+    val PC = Stageable(UInt(32 bits))
+    n0(PC) := 0x42
+    n0(PC, "true") := 0x42
+    n0(PC, 0x666) := 0xEE
+    val SOMETHING = n0.insert(myHardwareSignal) //This create a new Stageable
+    when(n1(SOMETHING) === 0xFFAA){ ... }
+    
+
+Also, there is an API to define nodes which are always valid / ready 
+
+.. list-table::
+   :header-rows: 1
+   :widths: 2 5
+
+   * - API
+     - Description
+   * - node.setAlwaysValid()
+     - Specify that the valid signal of the given node is always True. To use on the first node of a pipeline
```

Comment:
Is it really true that the first node in the pipeline would use `setAlwaysValid()`  At there description where you might not have a design that does this ?   Should a disabled via debugger of CPU IF stage modulate this state ?

Reply:
> Is it really true that the first node in the pipeline would use setAlwaysValid()

It can, it is equivalent to always doing node.valid := True

> At there description where you might not have a design that does this ?

When a design isn't always active, or when it is starving data (bubble)

> Should a disabled via debugger of CPU IF stage modulate this state ?

a disabled ?